src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +387,288 @@ public static void EnsureSocketsAreInitialized()
+        }
+
+        #region GetAddrInfoAsync Helper Classes
+
+        private sealed unsafe class GetAddrInfoExState : IDisposable
+        {
+            private IntPtr _context;
+            private int _callbackStartedOrCanceled;
+            private DnsResolveAsyncResult _asyncResult;
+
+            public string HostAddress { get; }
+
+            public GetAddrInfoExState(string hostAddress, DnsResolveAsyncResult asyncResult)
+            {
+                HostAddress = hostAddress;
+                _asyncResult = asyncResult;
+            }
+
+            public void SetQueryContext(IntPtr context)
+            {
+                _context = context;
+            }
+
+            public bool SetCallbackStartedOrCanceled()
+            {
+                return Interlocked.Exchange(ref _callbackStartedOrCanceled, 1) == 0;
+            }
+
+            public void CompleteAsyncResult(object o)
+            {
+                // We don't want to expose the GetAddrInfoEx callback thread to user code.
+
+                var state = Tuple.Create(_asyncResult, o);
+
+                Task.Factory.StartNew(s =>
+                {
+                    var t = (Tuple<DnsResolveAsyncResult, object>)s;
+                    t.Item1.InvokeCallback(t.Item2);
+                }, state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
+            }
+
+            ~GetAddrInfoExState()
+            {
+                // The finalizer should only be called when the AppDomain is unloading while there was a pending operation.
+                // Calling GetAddrInfoExCancel will invoke the callback inline with WSA_E_CANCELLED error code.
+
+                // If the callback is already invoked, let the traditional Dispose clear the resources.
+                // This should not be possible, because the finalizer would not have been called if the object was still alive.
+                // This is an extra protection in case the object was somehow resurrected.
+                if (!SetCallbackStartedOrCanceled())
+                    return;
+
+                ReleaseResources(cancelQuery: true);
+            }
+
+            public void Dispose()
+            {
+                GC.SuppressFinalize(this);
+                ReleaseResources(cancelQuery: false);
+            }
+
+            private void ReleaseResources(bool cancelQuery)
+            {
+                IntPtr ptr = Interlocked.Exchange(ref _context, IntPtr.Zero);
+
+                if (ptr == IntPtr.Zero)
+                    return;
+
+                var context = (GetAddrInfoExContext*)ptr;
+
+                if (cancelQuery)
+                    context->CancelQuery();
+
+                GetAddrInfoExContext.FreeContext(context);
+            }
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct GetAddrInfoExContext
+        {
+            private static readonly int Size = Marshal.SizeOf<GetAddrInfoExContext>();
```

Comment:
@stephentoub I think it's guaranteed to be the same since GetAddrInfoExContext is blittable. But do you prefer if I put `private static readonly int Size= Unsafe.SizeOf<GetAddrInfoExContext>();` instead?

Reply:
I think what you have is ok. Thanks.

Reply:
`Marshal.SizeOf` is slow expensive API. `sizeof(GetAddrInfoExContext)` would be a better here. I do not think you need to use `Unsafe.SizeOf`.